### PR TITLE
Fix ped plugin

### PR DESCRIPTION
**Pull request by jwillmartin**
# PR Details
## Description
These changes address a memory leak in the WebService/PSM functionality of the plugin. It also allows a user to switch between using FLIR(WebSocket) and PSM(WebService) without crashing or generating segmentation faults. 

## Related Issue

#497 

## Motivation and Context

Users of this plugin have noticed memory loss and error code generation while using this plugin. Users have also requested a fix for the #497 issue.

## How Has This Been Tested?
Tested using a script to send 1000 PSM XMLs to the plugin. No messages were lost. 
Tested using local FLIR camera to ensure functionality was maintained. 

Switched between DataProviders (PSM/FLIR) in various combinations to ensure threads were closed or joined properly. 

## Types of changes

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/v2i-hub/CommandPlugin/src/CommandPlugin.cpp [hunk 1]
```diff
@@ -633,7 +633,7 @@ int CommandPlugin::WSCallbackBASE64(
 											argsList[arg.first] = arg.second.data();
 										}
 									}
-									catch (exception argsEx)
+									catch (exception &argsEx)
```
Comment (paulbourelly999) [inline src/v2i-hub/CommandPlugin/src/CommandPlugin.cpp:636]:
```cpp
catch(const std::exception &argEx)
```

Also where you just address a code smell or why are there changes to this file at all.

Comment (jwillmartin) [reply src/v2i-hub/CommandPlugin/src/CommandPlugin.cpp]: This was just to address a code smell

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp [hunk 2]
```diff
@@ -17,124 +21,145 @@ namespace PedestrianPlugin
...
-	// Catch parse exceptions
-    try {
-	    BroadcastPsm(psmMsgdef);
-		writeResponse(QHttpEngine::Socket::Created, socket);
-	}
-	catch(const J2735Exception &e) {
-        PLOG(logERROR) << "Error parsing file: " << e.what() << std::endl;
-		writeResponse(QHttpEngine::Socket::BadRequest, socket);
-	}
+	if (_signSimClient != nullptr)
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:26]: Is this UDPClient ever used. If not lets delete it.

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp [hunk 3]
```diff
@@ -17,124 +21,145 @@ namespace PedestrianPlugin
...
+	// Catch parse exceptions
+    try {
+		for(const auto& psm_s: psmSL)
+		{
+			BroadcastPsm(psm_s);
+			socket->setStatusCode(QHttpEngine::Socket::Created);
+		}
+	}
+	catch(const J2735Exception &e) {
+        PLOG(logERROR) << "Error parsing file: " << e.what();
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:119]: **NIT** : Is the indentation off in this file or is github just doing funny stuff.

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp [hunk 4]
```diff
@@ -143,54 +168,60 @@ void PedestrianPlugin::UpdateConfigSettings()
...
 	// (e.g. using std::atomic, std::mutex, etc.).
 
-	int instance;
 	std::lock_guard<mutex> lock(_cfgLock);
-
-	GetConfigValue<string>("WebServiceIP",webip);
-	GetConfigValue<uint16_t>("WebServicePort",webport);
-	GetConfigValue<string>("WebSocketHost",webSocketIP);
-	GetConfigValue<string>("WebSocketPort",webSocketURLExt);
+	GetConfigValue<std::string>("WebServiceIP", webip);
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:172]: **NIT** Just an FYI. This works fine but there is an overloaded version of this method you can directly pass the mutex to, which avoids having to created the scoped `std::lock_guard`. See https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/4847aa7fdd30bf31f013cfe74fc3920d302fffe6/src/v2i-hub/SpatPlugin/src/SpatPlugin.cpp#L44C1-L45C1.

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp [hunk 5]
```diff
@@ -143,54 +168,60 @@ void PedestrianPlugin::UpdateConfigSettings()
...
-			PLOG(logDEBUG) << "Thread started!!: " << std::endl;
-			
-			webthread.detach(); // wait for the thread to finish
-
-			std::thread xmlThread(&PedestrianPlugin::checkXML,this);
-			PLOG(logDEBUG) << "XML Thread started!!: " << std::endl;
-			
-			xmlThread.detach(); // wait for the thread to finish
+    {
+		PLOG(logDEBUG) << "Before creating websocket to: " << webSocketIP.c_str() <<  " on port: " << webSocketURLExt.c_str();
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:185]: Can we remove this log statement

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp [hunk 6]
```diff
@@ -143,54 +168,60 @@ void PedestrianPlugin::UpdateConfigSettings()
...
+        StopWebSocket();
+        if (!runningWebService)
+        {
+			PLOG(logDEBUG) << "Starting WebService Thread";
+            std::thread webServiceThread(&PedestrianPlugin::StartWebService, this);
+			PLOG(logDEBUG) << "WebService Thread started";
+			webServiceThread.join(); // wait for the thread to finish
+        }
+    }
+	else
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:213]: Can we have some sort of WARNING log to indicate that the dataprovider configuration is misconfigured. Otherwise someone could have a typo and have no indication why the PSM plugin is not working as intended. Ex
```
PLOG (logWARNING) << "Invalid configured data provider. Valid values include FLIR and PSM!";
```

File: src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp [hunk 7]
```diff
@@ -62,51 +51,56 @@ namespace PedestrianPlugin
...
-};
-std::mutex _cfgLock;
+	bool runningWebSocket = false;
+    bool runningWebService = false;
+
+	// The io_context is required for all I/O
+    net::io_context ioc;
 
+	// API URL to accept PSM XML
+	const QString PSM_Receive = "";
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp:103]: This is never updated. Is this intentional. We should either set a fixed value for the endpoint or not have it as a const

File: src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp [hunk 8]
```diff
@@ -62,51 +51,56 @@ namespace PedestrianPlugin
...
+	
 	J2735MessageFactory factory;
 	
+	std::mutex _cfgLock;
+	
+	uint16_t webport;
+	std::string webip; 
+	std::string webSocketIP;
+	std::string webSocketURLExt;
+	int instance;
```
Comment (paulbourelly999) [inline src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp:90]: This variable is set but never used as far as I can tell. If I am correct lets remove it.

**Review by paulbourelly999 — CHANGES_REQUESTED** [review]
I left several small comments. For ones labeled **NIT** please address or don't address these as you see fit. For all others please address them.

File: ext/build.sh [hunk 9]
```diff
@@ -72,3 +72,11 @@ make -j${numCPU}
 make install
 popd
 
+# pugixml
+pushd /tmp
+git clone https://github.com/zeux/pugixml.git
```
[the remark this answers is not in the document]
Comment (jwillmartin) [reply ext/build.sh:77]: Good call, changed to v1.14

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp (hunk 2 above)
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:26]: It's not, removed.

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp (hunk 5 above)
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:185]: Removed

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp (hunk 6 above)
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:213]: Updated

File: src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp (hunk 7 above)
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp:103]: Yeah, this QObjectHandler requires a URL extension, but we don't want to add anything because we're using POST. For example, we want to keep 127.0.0.1:9000/ instead of 127.0.0.1:9000/someExtension.

Would you prefer to have empty quotes in the registerMethod or use this empty const? Thought this looked better, but it isn't needed.

File: src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp (hunk 8 above)
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp:90]: Not used, removed

File: src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwillmartin) [reply src/v2i-hub/PedestrianPlugin/src/PedestrianPlugin.cpp:119]: Updated

File: src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp (hunk 7 above)
Comment (paulbourelly999) [reply src/v2i-hub/PedestrianPlugin/src/include/PedestrianPlugin.hpp:132]: Would we break anything if we required some extension. It is probably better practice like /psm or /pedestriansafetymessage

**Review by paulbourelly999 — APPROVED** [review]
